rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,391 @@
+
+The metadata file currently only contains the `$cas` hashes of each runtime dependency, sorted alphabetically, and separated by single newlines. If the file does not match this format exactly, it is invalid.
+
+Other objective metadata that could be useful, such as late binding information, could be placed in this file too, perhaps as a TOML file that begins after the first double newline. This is to be determined.
+
+### Trust DB
+
+The Trust DB contains subjective metadata for `$out` -> `$cas` mappings. Each entry contains:
+
+- `$cas`
+- an array of `$out`
+- name (including version)
+- size
+- array of dependency `$cas`
+- optional:
+  - description
+  - build timestamp
+  - builder
+  - custom metadata, like one or more git repo commit hashes
+
+With this information, a user can quickly find `$cas` entries to install that match a name or description. `nix-build` can find a `$cas` by `$out`.
+
+For a given `$cas`, there can be many entries, one for each trusted source. This can be handled by having one SQLite DB per source (including localhost), and having an order of precedence.
+
+### Sharing the Nix Store
+
+Since the Nix Store (minus supporting directories) contains only self-validating paths, it can be shared "infinitely", only limited by:
+
+- disk space
+- network performance
+- confidence around hash collision attacks
+- confidence around writers corrupting paths without detection
+
+The installation step only involves moving a proposed path from `.prepare` to `.stage`, so no further communication is necessary with the Store daemon.
+
+For single-user installs, the Store can trivially be maintained by the Nix tools, and converting to multi-user is only a matter of changing the permissions.
+
+It would even be possible to use FUSE to automatically download any paths that are referenced in the Store, hanging the I/O request while it's being downloaded.
+
+### Store Daemon
+
+Optionally, a daemon can maintain the Store. In this case, it is recommended be the only user with write access. It performs installations, verifications and garbage collection, described below.
+
+### Preparing
+
+`$cas` entries are prepared by building them in a path that has the same total length as its final `$cas` entry. This means they can be built practically anywhere, in `/tmp`, in a home directory, in the `.prepare` directory in the Store, etc. It is recommended to make a part of the path a unique-per-build string.
+To make sure only the build's own references need rewriting, it is recommended to build using only `$cas` entries as dependencies, instead of relying on rewriting paths.
```

Comment:
```suggestion
To make sure only the build's self references need rewriting, it is recommended to build using only `$cas` entries as dependencies, instead of relying on rewriting paths.
```